File: ab-cloud-verification/r/ab_cloud_verify_en.R
# ==============================================================================
# AB-Cloud Verification Suite — R Implementation (English)
# ==============================================================================
# English-only version. For bilingual, use ab_cloud_verify.R
# ==============================================================================

MSG_EN <- list(
  header       = "AB-Cloud Verification Suite — R",
  separator    = "──────────────────────────────────────────────────────",
  loading      = "Loading zeros from: %s",
  loaded       = "Loaded %d zeros from %s",
  obj1_title   = "Objection 1: b(N) Convergence",
  obj1_desc    = "b(N) = (1/N) * Σ|γ_k - γ̃_k|, Gram points via Lambert W",
  obj1_n       = "N",
  obj1_bN      = "b(N)",
  obj1_status  = "Status",
  obj1_converge = "CONVERGING — b(N) → 0 supports AB-Cloud",
  obj1_stable  = "STABLE — b(N) near zero, AB-Cloud consistent",
  obj1_diverge = "DIVERGING — b(N) not → 0, objection upheld",
  obj2_title   = "Objection 2: GUE Spacing KS Test",
  obj2_desc    = "s_k = (γ_{k+1}-γ_k)·log(γ_k/2π)/(2π), vs p(s)=(πs/2)·exp(-πs²/4)",
  obj2_stat    = "D-statistic",
  obj2_pval    = "p-value",
  obj2_result  = "Result",
  obj2_pass    = "PASS — GUE spacing confirmed (p > 0.05)",
  obj2_fail    = "FAIL — GUE spacing rejected (p ≤ 0.05)",
  obj3_title   = "Objection 3: Large-T Decay Slope",
  obj3_desc    = "Linear regression of log|γ_k - γ̃_k| vs log(γ_k), expect slope ≈ -0.5",
  obj3_slope   = "Slope",
  obj3_stderr  = "Std Error",
  obj3_target  = "Target",
  obj3_result  = "Result",
  obj3_pass    = "PASS — Slope ≈ -0.5, AB-Cloud decay confirmed",
  obj3_fail    = "FAIL — Slope deviates from -0.5",
  no_data      = "ERROR: No zeros loaded. Check data directory.",
  done         = "Verification complete."
)

# --- Lambert W (principal branch) via Halley's method -------------------------

lambert_W0_en <- function(x, tol = 1e-12, max_iter = 50) {
  if (x == 0) return(0)
  w <- if (x > 1) log(x) - log(log(x)) else x
  for (i in seq_len(max_iter)) {
    ew  <- exp(w)
    f   <- w * ew - x
    fp  <- ew * (1 + w)
    fpp <- ew * (2 + w)
    w   <- w - (2 * f * fp) / (2 * fp * fp - f * fpp)
    if (abs(f) < tol * abs(x + 1)) break
  }
  return(w)
}

# --- Gram point computation via Lambert W ------------------------------------

gram_point_en <- function(n) {
  if (n <= 0) return(0)
  val <- 2 * pi * n / lambert_W0_en(n / exp(1))
  for (iter in 1:3) {
    theta  <- 0.5 * val * log(val / (2 * pi)) - 0.5 * val - pi / 8
    dtheta <- 0.5 * log(val / (2 * pi))
    val    <- val + (pi * n - theta) / dtheta
  }
  return(val)
}

gram_points_vec_en <- function(n_vec) sapply(n_vec, gram_point_en)

# --- GUE spacing CDF (Wigner surmise) ----------------------------------------

gue_cdf_en <- function(s) 1 - exp(-pi * s^2 / 4)

# --- Load zeros from data files ----------------------------------------------

load_zeros_en <- function(dataDir, count = 0, source = "auto") {
  files <- list(
    zeta_zeros_50000      = file.path(dataDir, "zeta_zeros_50000.txt"),
    zeta_zeros_500k       = file.path(dataDir, "zeta_zeros_500k_odlyzko.txt"),
    zeta_zeros_2M         = file.path(dataDir, "zeta_zeros_2M_odlyzko.txt"),
    zeta_zeros_highT      = file.path(dataDir, "zeta_zeros_highT_blocks.txt"),
    zeros6                = file.path(dataDir, "zeros6.txt"),
    zeta_zeros_50000_csv  = file.path(dataDir, "zeta_zeros_50000.csv")
  )
  if (source != "auto" && source %in% names(files)) {
    selected <- files[[source]]
  } else {
    selected <- NULL
    for (nm in names(files)) {
      if (file.exists(files[[nm]])) {
        selected <- files[[nm]]
        if (count > 0 && count <= 50000 && nm == "zeta_zeros_50000") break
      }
    }
  }
  if (is.null(selected) || !file.exists(selected))
    stop("No data file found in: ", dataDir)

  cat(sprintf(MSG_EN$loading, selected), "\n")
  ext <- tools::file_ext(selected)
  zeros <- if (ext == "csv") {
    df <- read.csv(selected, comment.char = "#", header = FALSE)
    as.numeric(df[[1]])
  } else {
    lines <- readLines(selected, warn = FALSE)
    lines <- lines[!grepl("^\\s*#", lines) & nzchar(trimws(lines))]
    vals  <- as.numeric(trimws(lines))
    vals[!is.na(vals)]
  }
  if (count > 0 && count < length(zeros)) zeros <- zeros[1:count]
  cat(sprintf(MSG_EN$loaded, length(zeros), basename(selected)), "\n")
  return(zeros)
}

# --- Objection 1: b(N) Convergence ------------------------------------------

objection_1_en <- function(zeros) {
  M <- MSG_EN; N <- length(zeros)
  cat("\n", M$separator, "\n", M$obj1_title, "\n", M$obj1_desc, "\n", M$separator, "\n\n")
  checkpoints <- c(100, 500, 1000, 5000, 10000, 50000, 100000, 500000, 1000000)
  checkpoints <- checkpoints[checkpoints <= N]
  results <- data.frame(N = integer(), bN = numeric(), Status = character(), stringsAsFactors = FALSE)
  prev_bN <- NA
  for (cp in checkpoints) {
    idx  <- 1:cp
    gram <- gram_points_vec_en(idx)
    bN   <- mean(abs(zeros[idx] - gram))
    status <- if (is.na(prev_bN)) "—" else if (bN < prev_bN * 1.05) "↓" else "↑"
    results <- rbind(results, data.frame(N = cp, bN = bN, Status = status, stringsAsFactors = FALSE))
    prev_bN <- bN
  }
  cat(sprintf("%10s  %14s  %8s\n", M$obj1_n, M$obj1_bN, M$obj1_status))
  cat(sprintf("%10s  %14s  %8s\n", "──────────", "──────────────", "────────"))
  for (i in seq_len(nrow(results)))
    cat(sprintf("%10d  %14.8f  %8s\n", results$N[i], results$bN[i], results$Status[i]))
  final_bN <- results$bN[nrow(results)]
  verdict <- if (final_bN < 0.01) M$obj1_converge else if (final_bN < 0.5) M$obj1_stable else M$obj1_diverge
  cat("\n", verdict, "\n")
  return(results)
}

# --- Objection 2: GUE Spacing KS Test ----------------------------------------

objection_2_en <- function(zeros) {
  M <- MSG_EN; N <- length(zeros)
  cat("\n", M$separator, "\n", M$obj2_title, "\n", M$obj2_desc, "\n", M$separator, "\n\n")
  gammas  <- zeros[1:(N - 1)]
  deltas  <- zeros[2:N] - zeros[1:(N - 1)]
  log_fac <- log(gammas / (2 * pi)) / (2 * pi)
  s       <- deltas * log_fac
  s       <- s[is.finite(s) & s > 0]
  ks_res  <- ks.test(s, gue_cdf_en)
  cat(sprintf("%18s: %.8f\n", M$obj2_stat, ks_res$statistic))
  cat(sprintf("%18s: %.6e\n", M$obj2_pval, ks_res$p.value))
  verdict <- if (ks_res$p.value > 0.05) M$obj2_pass else M$obj2_fail
  cat("\n", verdict, "\n")
  return(list(D = ks_res$statistic, p.value = ks_res$p.value))
}

# --- Objection 3: Large-T Decay Slope ----------------------------------------

objection_3_en <- function(zeros) {
  M <- MSG_EN; N <- length(zeros)
  cat("\n", M$separator, "\n", M$obj3_title, "\n", M$obj3_desc, "\n", M$separator, "\n\n")
  start <- max(1, floor(N * 0.5))
  idx   <- start:N
  gram  <- gram_points_vec_en(idx)
  dev   <- abs(zeros[idx] - gram)
  valid <- is.finite(dev) & dev > 0
  log_gamma <- log(zeros[idx][valid])
  log_dev   <- log(dev[valid])
  fit    <- lm(log_dev ~ log_gamma)
  slope  <- coef(fit)[2]
  stderr <- summary(fit)$coefficients[2, 2]
  cat(sprintf("%18s: %.6f\n", M$obj3_slope, slope))
  cat(sprintf("%18s: %.6f\n", M$obj3_stderr, stderr))
  cat(sprintf("%18s: -0.5\n", M$obj3_target))
  verdict <- if (abs(slope - (-0.5)) < 0.15) M$obj3_pass else M$obj3_fail
  cat("\n", verdict, "\n")
  return(list(slope = slope, stderr = stderr))
}

# --- Main entry (English only) -----------------------------------------------

ab_cloud_verify_en <- function(dataDir = "../data", zeros = 0, source = "auto",
                                objection = "all") {
  M <- MSG_EN
  cat("\n", M$separator, "\n", M$header, "\n", M$separator, "\n")
  gammas <- load_zeros_en(dataDir, count = zeros, source = source)
  if (length(gammas) == 0) { cat(M$no_data, "\n"); return(invisible(NULL)) }
  results <- list()
  if (objection %in% c("all", "1")) results$obj1 <- objection_1_en(gammas)
  if (objection %in% c("all", "2")) results$obj2 <- objection_2_en(gammas)
  if (objection %in% c("all", "3")) results$obj3 <- objection_3_en(gammas)
  cat("\n", M$separator, "\n", M$done, "\n\n")
  return(invisible(results))
}

File: ab-cloud-verification/julia/ab_cloud_verify_en.jl
#!/usr/bin/env julia
# =============================================================================
# AB-CLOUD VERIFICATION SUITE — ENGLISH
# Idiomatic Julia · Type Annotations · Efficient
# =============================================================================
# Three objections against the Riemann Hypothesis:
#   Objection 1: b(N) → 0?  (Gram point deviation convergence)
#   Objection 2: GUE spacing KS test (level spacing distribution)
#   Objection 3: Large-T decay slope ≈ -0.5
# =============================================================================

using Printf

const PI::Float64     = 3.14159265358979323846
const TWO_PI::Float64 = 6.28318530717958647693

# --- Load zeros ---
function load_zeros(n_request::Int; source::String="auto")::Vector{Float64}
    filepath::String = if source != "auto"
        "../data/$source"
    elseif n_request ≤ 13661
        "../data/zeta_zeros_50000.txt"
    elseif n_request ≤ 500_000
        "../data/zeta_zeros_500k_odlyzko.txt"
    elseif n_request ≤ 2_000_000
        "../data/zeta_zeros_2M_odlyzko.txt"
    else
        "../data/zeros6.txt"
    end
    println("  Data file: $filepath")
    zeros::Vector{Float64} = Float64[]
    open(filepath, "r") do io
        for line in eachline(io)
            line = strip(line); isempty(line) && continue; startswith(line, '#') && continue
            try; v = parse(Float64, line); v > 0.0 && push!(zeros, v); catch; continue; end
        end
    end
    length(zeros) > n_request && resize!(zeros, n_request)
    return zeros
end

# --- Lambert W ---
function lambert_w0(x::Float64)::Float64
    x == 0.0 && return 0.0
    w::Float64 = log(max(x, 1e-30)); w > 0.0 && (w = log(w))
    for _ in 1:50
        ew = exp(w); f = w*ew - x; fp = ew*(w+1.0); fpp = ew*(w+2.0)
        delta = f / (fp - 0.5*f*fpp/fp); w -= delta
        abs(delta) < 1e-15*abs(w) && break
    end
    return w
end

gram_point(n::Int)::Float64 = TWO_PI * lambert_w0(n / exp(1.0))

# --- Objection 1 ---
function objection1(gammas::Vector{Float64})
    N = length(gammas); s = 0.0
    @inbounds for k in 1:N; s += abs(gammas[k] - gram_point(k)); end
    b_n = s / N; pass = b_n < 1.0
    println("\n=====================================================")
    println("  OBJECTION 1: b(N) Convergence")
    println("  b(N) = (1/N) * Sum|gamma_k - gamma~_k|, Gram pts via Lambert W")
    println("=====================================================")
    @printf("    N       = %d\n", N); @printf("    b(N)    = %.8e\n", b_n)
    println("    Pass:   $pass")
    return b_n, pass
end

# --- GUE helpers ---
gue_pdf(s::Float64)::Float64 = (PI*s/2.0) * exp(-PI*s^2/4.0)

function gue_cdf(s::Float64)::Float64
    s ≤ 0.0 && return 0.0
    nsteps = 200; h = s / nsteps; sum_val = 0.0
    for i in 0:nsteps
        x = i * h
        if i == 0 || i == nsteps; sum_val += gue_pdf(x)
        elseif isodd(i); sum_val += 4.0*gue_pdf(x)
        else; sum_val += 2.0*gue_pdf(x); end
    end
    return min(h/3.0*sum_val, 1.0)
end

function kolmogorov_pvalue(d::Float64, n::Float64)::Float64
    z = d*sqrt(n); pval = 0.0
    for k in -10:10; pval += (-1)^k * exp(-2.0*(2k*z+z)^2); end
    return clamp(pval, 0.0, 1.0)
end

# --- Objection 2 ---
function objection2(gammas::Vector{Float64})
    N = length(gammas); m = N - 1
    s = Vector{Float64}(undef, m)
    @inbounds for k in 1:m
        s[k] = (gammas[k+1] - gammas[k]) * log(gammas[k]/TWO_PI) / TWO_PI
    end
    sort!(s)
    ks_stat = 0.0
    @inbounds for k in 1:m
        d_plus = abs(k/m - gue_cdf(s[k])); d_plus > ks_stat && (ks_stat = d_plus)
    end
    ks_pval = kolmogorov_pvalue(ks_stat, Float64(m)); pass = ks_pval > 0.05
    println("\n=====================================================")
    println("  OBJECTION 2: GUE Spacing KS Test")
    println("  s_k = dg_k * log(g_k/2pi) / 2pi,  p(s) = (pi*s/2)*exp(-pi*s^2/4)")
    println("=====================================================")
    @printf("    Spacing count    = %d\n", m)
    @printf("    KS statistic     = %.8e\n", ks_stat)
    @printf("    p-value          = %.8e\n", ks_pval)
    println("    Pass (p>0.05):   $pass")
    return ks_stat, ks_pval, pass
end

# --- Objection 3 ---
function objection3(gammas::Vector{Float64})
    N = length(gammas); i_start = div(N,2)+1; m = N - i_start
    m < 10 && return 0.0, false
    x_mean = 0.0; y_mean = 0.0
    @inbounds for k in i_start:(N-1)
        x_mean += log(gammas[k]); y_mean += log(abs(gammas[k+1]-gammas[k]))
    end
    x_mean /= m; y_mean /= m
    sxx = 0.0; sxy = 0.0
    @inbounds for k in i_start:(N-1)
        xv = log(gammas[k]); yv = log(abs(gammas[k+1]-gammas[k]))
        sxx += (xv-x_mean)^2; sxy += (xv-x_mean)*(yv-y_mean)
    end
    slope = sxy/sxx; dev = abs(slope+0.5); pass = dev < 0.15
    println("\n=====================================================")
    println("  OBJECTION 3: Large-T Decay Slope")
    println("  log|dg_k| ~ slope * log(g_k),  expected slope = -0.5")
    println("=====================================================")
    @printf("    Regression pts  = %d\n", m)
    @printf("    Slope           = %.4f\n", slope)
    @printf("    Deviation       = %.4f\n", dev)
    println("    Pass:           $pass")
    return slope, pass
end

# --- Main ---
function main()
    n_zeros = 10000; source = "auto"; objection = 0
    args = ARGS; i = 1
    while i ≤ length(args)
        arg = args[i]
        if arg == "--zeros" && i+1 ≤ length(args); i += 1; n_zeros = parse(Int, args[i])
        elseif arg == "--source" && i+1 ≤ length(args); i += 1; source = args[i]
        elseif arg == "--objection" && i+1 ≤ length(args)
            i += 1; objection = args[i] == "all" ? 0 : parse(Int, args[i])
        end; i += 1
    end

    println(); println("  +=================================================+")
    println("  |  AB-CLOUD VERIFICATION — RIEMANN HYPOTHESIS     |")
    println("  |  Three objections: b(N), GUE KS, Large-T        |")
    println("  +=================================================+"); println()

    gammas = load_zeros(n_zeros; source=source)
    println("  Zeros loaded: $(length(gammas)) (gamma ordinates)")

    t = time()
    results = []
    if objection == 0 || objection == 1; push!(results, ("1:b(N)", objection1(gammas)...)); end
    if objection == 0 || objection == 2; push!(results, ("2:KS", objection2(gammas)[1], objection2(gammas)[3])); end
    if objection == 0 || objection == 3; push!(results, ("3:slope", objection3(gammas)...)); end

    println("\n  +----------+--------------------+--------+")
    println("  | Obj.     |    Value           | Status |")
    println("  +----------+--------------------+--------+")
    for (lbl, val, p) in results
        @printf("  | %-8s | %16.6e | %-6s |\n", lbl, val, p ? "PASS" : "FAIL")
    end
    println("  +----------+--------------------+--------+")
    @printf("  Total time: %.3f s\n", time() - t)
end

main()
